autogen/coding/base.py

```diff
@@ -0,0 +1,94 @@
+from typing import Any, Dict, List, Protocol, Union, runtime_checkable
+
+from pydantic import BaseModel, Field
+
+from ..agentchat.agent import LLMAgent
+
+__all__ = ("CodeBlock", "CodeResult", "CodeExtractor", "CodeExecutor")
+
+
+class CodeBlock(BaseModel):
+    """A class that represents a code block."""
+
+    code: str = Field(description="The code to execute.")
+
+    language: str = Field(description="The language of the code.")
+
+
+class CodeResult(BaseModel):
+    """A class that represents the result of a code execution."""
+
+    exit_code: int = Field(description="The exit code of the code execution.")
+
+    output: str = Field(description="The output of the code execution.")
+
+
+class CodeExtractor(Protocol):
```

Comment:
Why do we need `CodeExtractor` as a separate class, rather than having the `extract_code_blocks` method within the `CodeExecutor`? 

Is that because one executor might have multiple extractor?

Reply:
Got it... It is because we can have different extractor (e.g., extract from markdown, json, etc.) for the same executor (e.g., Python).